File: pyth/longest common subsq/lcs.py

def lcs(str1, str2, len1, len2):
    global dp
    if len1 <= 0 or len2 <= 0:
        return 0
    else:
        if dp[len1-1][len2-1] == -1:
            if str1[len1-1] == str2[len2-1]:
                dp[len1-1][len2-1] = 1 + lcs(str1, str2, len1-1, len2-1)
            else:
                temp1 = lcs(str1, str2, len1-1, len2)
                temp2 = lcs(str1, str2, len1, len2-1)
                dp[len1-1][len2-1] = max(temp1, temp2)
        return dp[len1-1][len2-1]
    
n = 0               #taking input
while n<2:
    if n==0:
        str1 = input()
    else:
        str2 = input()
    n += 1


rows = len(str1)
cols = len(str2)
dp = [[-1]*(cols+1) for j in range(rows+1)]
k = lcs(str1, str2, rows, cols)
print(k)

ans = ""
i = len(str1)-1
j = len(str2)-1

while(i>=0  and  j>=0):    #getting the longest common subsequence
    if str1[i] == str2[j]:
        ans = str1[i] + ans
        i -= 1
        j -= 1
    else:
        if dp[i-1][j] > dp[i][j-1]:
            i -= 1
        else:
            j -= 1
print (ans)            
    
    
    

File: lcs.cpp
#include<iostream>   //dynamic programming approach to longest common subsequence problem
#include<string>
using namespace std;
int dp[100][100];


int rec(string str1, string str2, int sz1, int sz2)
{
    if(sz1 <= 0 || sz2 <= 0)
        return 0;
    else
    {
        if(dp[sz1-1][sz2-1] == -1)
        {
            if(str1[sz1-1] == str2[sz2-1])          
                    dp[sz1-1][sz2-1] = 1+rec(str1, str2, sz1-1, sz2-1);
            else 
            {

                int temp1 = rec(str1, str2, sz1-1, sz2);
                int temp2 = rec(str1, str2, sz1, sz2-1);
                dp[sz1-1][sz2-1] = max(temp1, temp2);
            }
        }
        return dp[sz1-1][sz2-1];
    }
}

int main()
{
    string str1, str2;
    cout << "Enter the string to be compared :\n";
    cin >> str1 >> str2;
    for(int i=0; i<str1.size(); i++)
        for(int j=0; j<str2.size(); j++)
            dp[i][j] = -1 ;   
    int len = rec(str1, str2, str1.size(), str2.size());
    cout <<"Length of longest common sbusequence is: " << len << "\n";

    // for(int i=0; i<str1.size(); i++)
    // {
    //     for(int j=0; j<str2.size(); j++)
    //         cout << dp[i][j] << " ";
    //     cout << "\n";
    // }    

    string ans;
    int i = str1.size()-1; int j = str2.size()-1;
    while(i >= 0  && j >= 0)
        if(str1[i] == str2[j])
        {
            ans += str1[i];
            i--; j--;
        }   
        else
        {
            if(dp[i-1][j] > dp[i][j-1])
                i--;
            else 
                j--;
        }
    cout << "Subsequence is: ";
    for(int k=len-1; k>=0; k--)
        cout << ans[k];
    cout << "\n";
}